File: SOLID/Interface.cpp
#include <iostream>
using namespace std;

//Seria incorrecto
class Dispositivo{
    virtual void apagar() = 0;
    virtual void reiniciar() = 0;
    virtual void tomarFoto() = 0;
};

//Lo correcto es crear interfaces con menos métodos
class Electronico {
public:
    virtual void apagar() = 0;
    virtual void reiniciar() = 0;
};

class Camara {
public:
    virtual void tomarFoto() = 0;
};

class Computadora : public Electronico {
public:
    void apagar() override {
        cout << "Apagando computadora." << endl;
    }

    void reiniciar() override {
        cout << "Reiniciando computadora." << endl;
    }
};

class Celular : public Electronico, public Camara{
public:
    void apagar() override {
        cout << "Apagando teléfono." << endl;
    }

    void reiniciar() override {
        cout << "Reiniciando teléfono." << endl;
    }

    void tomarFoto() override {
        cout << "Tomando foto con el teléfono." << endl;
    }
};

int main() {
    Computadora dispositivo_digital;
    dispositivo_digital.apagar();
    dispositivo_digital.reiniciar();
    Celular otro;
    otro.apagar();
    otro.reiniciar();

    return 0;
}

File: SOLID/Interface.py
from abc import ABC, abstractmethod

#Seria incorrecto
class Dispositivo:
    @abstractmethod
    def apagar(self):
        pass
    @abstractmethod
    def reiniciar(self):
        pass
    @abstractmethod
    def tomarFoto(self):
        pass


#Lo correcto es crear interfaces con menos métodos
class Electronico(ABC):
    @abstractmethod
    def apagar(self):
        pass
    @abstractmethod
    def reiniciar(self):
        pass

class Camara(ABC):
    @abstractmethod
    def tomarFoto(self):
        pass

class Computadora(Electronico):
    def apagar(self):
        print("Apagando computadora.")

    def reiniciar(self):
        print("Reiniciando computadora.")


class Celular(Electronico, Camara):
    def apagar(self):
        print("Apagando teléfono.")
    def reiniciar(self):
        print("Reiniciando teléfono.")
    def tomarFoto(self):
        print("Tomando foto con el teléfono.")

if __name__ == "__main__":
    dispositivo_digital = Computadora()
    dispositivo_digital.apagar()
    dispositivo_digital.reiniciar()

    otro = Celular()
    otro.apagar();
    otro.reiniciar();
